enarx-keep-sgx-shim/src/heap.rs

```diff
@@ -0,0 +1,198 @@
+            allocated,
+            pages,
+        }
+    }
+
+    const fn idx_bit(page: usize) -> (usize, usize) {
+        let idx = page / (size_of::<Word>() * 8);
+        let bit = page % (size_of::<Word>() * 8);
+        (idx, bit)
+    }
+
+    fn is_allocated(&self, page: usize) -> bool {
+        let (idx, bit) = Self::idx_bit(page);
+        self.allocated[idx] & (1 << bit) != 0
+    }
+
+    fn is_allocated_range(&self, range: core::ops::Range<usize>) -> bool {
+        for i in range {
+            if self.is_allocated(i) {
+                return true;
+            }
+        }
+
+        false
+    }
+
+    fn allocate(&mut self, page: usize) {
+        let (idx, bit) = Self::idx_bit(page);
+        self.allocated[idx] |= 1 << bit;
+    }
+
+    fn deallocate(&mut self, page: usize) {
+        let (idx, bit) = Self::idx_bit(page);
+        self.allocated[idx] &= !(1 << bit);
+    }
+
+    fn offset(&self, addr: usize) -> Option<usize> {
+        let base = self.pages.as_ptr() as usize;
+        let ceil = base + self.pages.len() * Page::size();
+
+        if base <= addr && addr < ceil {
+            return Some(addr - base);
+        }
+
+        None
+    }
+
+    fn offset_page_down(&self, addr: usize) -> Option<usize> {
+        let off = self.offset(addr)?;
+        Some(off / Page::size())
+    }
+
+    fn offset_page_up(&self, addr: usize) -> Option<usize> {
+        self.offset_page_down(addr + Page::size() - 1)
+    }
+
+    pub fn brk(&mut self, brk: usize) -> usize {
+        let old = self.offset_page_up(self.metadata.brk.end).unwrap();
+        let new = match self.offset_page_up(brk) {
+            Some(page) => page,
+            None => return self.metadata.brk.end,
+        };
+
+        if old < new {
+            for page in old..new {
+                if self.is_allocated(page) {
+                    return self.metadata.brk.end;
+                }
+            }
+
+            for page in old..new {
+                self.allocate(page);
+            }
+        } else {
+            for page in old..new {
+                self.deallocate(page);
+            }
+        }
+
+        self.metadata.brk.end = brk;
+        brk
+    }
+
+    pub fn mmap(
+        &mut self,
+        addr: usize,
+        length: usize,
+        prot: usize,
+        flags: usize,
+        fd: isize,
+        offset: usize,
+    ) -> usize {
+        const RWX: u64 = nolibc::PROT_READ | nolibc::PROT_WRITE | nolibc::PROT_EXEC;
+        const PA: u64 = nolibc::MAP_PRIVATE | nolibc::MAP_ANONYMOUS;
+
+        let prot = prot as u64 & !RWX;
+        if addr != 0 || fd != -1 || offset != 0 || prot != 0 || flags as u64 != PA {
+            return ErrNo::EINVAL.into_syscall() as _;
+        }
+
+        // The number of pages we need for the given length.
+        let pages = (length + Page::size() - 1) / Page::size();
+
+        // Find the brk page offset.
+        let brk = self.offset_page_up(self.metadata.brk.end).unwrap();
```

Comment:
If we trip the unwrap, is it a failure case we can handle and recover from in here?

Reply:
We need a better solution here, I agree. A failed `unwrap()` here causes a deadlock (infinite loop in the panic handler). I know, because I hit it.